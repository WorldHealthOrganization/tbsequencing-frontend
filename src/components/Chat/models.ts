export interface IUserMessage {
  email: string,
  username: string,
  firstName: string,
  lastName: string
}
