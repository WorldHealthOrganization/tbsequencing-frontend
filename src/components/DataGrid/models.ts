export interface IColumn<T> {
  name: keyof T;
  header: string;
}

export enum GridType {
  Gene = 'gene',
  Drug = 'drug',
}

export enum ScrollType {
  Infinite = 'infinite',
  Regular = 'regular',
}

export type SortOrder = 'asc' | 'desc' | undefined;

export interface ISortState {
  colIndex?: number
  order?: SortOrder
  colID?: string
}

export interface IColumnFilter {
  id: string;
  value: number | string;
}

export interface BaseColumns {
  variantName: string;
  variantGrade: string;
  consequence: string;
  globalFrequency: string;
  totalCounts: number;
  resistantCount: number;
  susceptbleCount: number;
  intermediateCount: number;
  nucleodicAnnName: string;
  proteicAnnName: string;
  drug: number;
  drugName: string;
}

export interface DrugTableColumns extends BaseColumns {
  geneName: string;
}

export interface GeneTableColumns extends BaseColumns {
  drug: number;
}
