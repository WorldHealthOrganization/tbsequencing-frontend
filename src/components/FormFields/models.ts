import {
  ControllerRenderProps, FieldError, FieldName, FieldValues,
} from 'react-hook-form';
import { InputProps } from '@mui/material';
import { AbstractIField, CustomInputType } from '../../features/auth/components/registration/types';

export interface IBaseControlProps<IInput> extends
  AbstractIField<IInput>,
  Pick<ControllerRenderProps<FieldValues, FieldName<IInput>>, 'value' | 'onChange' | 'onBlur' > {
  isError: boolean,
  error: FieldError,
  fullWidth?: boolean,
  required?: boolean,
}

export interface ICheckboxProps<IInput> extends IBaseControlProps<IInput> {

}

export interface ITextAreaProps<IInput> extends IBaseControlProps<IInput> {

}

export interface ITextControlProps<IInput> extends IBaseControlProps<IInput> {
  type: CustomInputType
  InputProps?: Partial<InputProps>
}

export interface ISwitchControlProps<IInput> extends IBaseControlProps<IInput> {
  isActive?: boolean;
}

export interface ISelectControlProps<IInput> extends IBaseControlProps<IInput> {

}
