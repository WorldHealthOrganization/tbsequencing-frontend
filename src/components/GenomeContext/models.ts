import { IGene } from '../../services/genesApi/models';

export interface IGenomeData extends IGene {
  isSelected: boolean;
  y: number;
}
