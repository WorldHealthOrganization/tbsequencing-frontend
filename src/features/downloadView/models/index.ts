import { IDrug } from '../../../services/drugsApi/models';

export interface IExtendedDrug extends IDrug {
  isChecked: boolean;
}
