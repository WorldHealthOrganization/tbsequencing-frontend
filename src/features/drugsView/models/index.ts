export enum DrugStatus {
  Susceptible = 'susceptible',
  Resistant = 'resistant',
  Intermediate = 'intermediate',
  RatioSusceptible = 'ratioSusceptible',
  RatioResistant = 'ratioResistant',
  RatioIntermediate = 'ratioIntermediate',
}

export enum ResistantType {
  Phenotypic = 'Pheno',
  Genotypic = 'Geno',
}

export enum ChartDataType {
  Ratio = 'ratio',
  Count = 'count',
}

export enum Regions {
  ALL = 'ALL_REGIONS',
  AFRICA = 'AFR',
  EUROPE = 'EUR',
  AMERICA = 'AMR',
  EASTERN_MEDITERRANEAN = 'EMR',
  WESTERN_PACIFIC = 'WPR',
  SOUTH_EAST_ASIA = 'SEA',
}
