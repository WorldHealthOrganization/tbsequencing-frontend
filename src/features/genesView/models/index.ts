import { skipToken } from '@reduxjs/toolkit/query';
import React from 'react';
import { IGene, IVariant } from '../../../services/genesApi/models';

export type AutocompleteParams = typeof skipToken | { search?: string };

export interface IGenomeContextWrapperProps {
  data?: IGene[];
  isLoading?: boolean;
  selectedGene?: IGenesOverview;
  minDomain?: number;
  maxDomain?: number;
}

export interface IGenomeContextProps extends IGenomeContextWrapperProps {
  width: number;
}

export interface IVariantLocal extends IVariant {
  pos: number;
}

export interface ISearchResult {
  geneName: string;
  geneDbCrossrefId: number;
  locusTag: string;
}

export interface ISearchCount extends ISearchResult {
  counter: number;
}

export interface ISearchRecent extends ISearchResult {
  counter: string;
  date: Date | string;
}

export interface IGenesOverview extends IGene {
  geneDescription: string;
  geneType: string;
  ncbiId: string;
  proteinLength: number;
}

export interface IGenesOverviewTableData {
  title: string;
  value: string | number | null;
  element?: React.ReactNode;
}

export interface IGenesOverviewResponse {
  genes: IGene;
  genesOverview: IGenesOverview
}
