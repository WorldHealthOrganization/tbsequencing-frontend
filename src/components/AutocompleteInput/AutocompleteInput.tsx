/** @jsxImportSource @emotion/react */
import React from 'react';
import {
  Autocomplete, SxProps, TextField,
} from '@mui/material';
import { IGene } from '../../services/genesApi/models';
import { ISearchCount } from '../../features/genesView/models';
import { notNil } from '../../utils/arrays';

interface IAutocompleteProps {
  isLoading: boolean;
  style: SxProps;
  value?: string;
  options: IGene[] | ISearchCount[];
  onChange: (event: React.SyntheticEvent, value: string) => void;
  onSelectionChanged: (event: React.SyntheticEvent, value: IGene | null) => void;
  selectedValue: IGene | null;
}

const findProperLabel = (input: string, option: IGene) => {
  const keyOptions = Object.values(option);
  const label = keyOptions
    .filter(notNil)
    .map((filterOption) => filterOption.toString())
    .find((keyOption) => keyOption.toString().toLocaleLowerCase()
      .includes(input.toLocaleLowerCase()));

  return label ? label.toString() : option.geneName || option.locusTag;
};

export const AutocompleteInput = ({
  style, options, onChange, onSelectionChanged, selectedValue, isLoading, value,
}: IAutocompleteProps) => {
  const isEmpty = !value;

  return (
    <Autocomplete
      data-testid="autocomplete"
      renderOption={(props, option) => {
        if (isEmpty) {
          return (
            <li
              {...props}
              key={option.geneDbCrossrefId}
            >
              {option.geneName}
            </li>
          );
        }

        const labelName = findProperLabel(value, option);

        return <li {...props} key={option.geneDbCrossrefId}>{labelName}</li>;
      }}
      getOptionLabel={(option) => (isEmpty ? option.geneName : findProperLabel(value, option))}
      renderInput={
        (params) => (
          <TextField
            sx={style}
            {...params}
            placeholder="Search for gene..."
          />
        )
      }
      onChange={onSelectionChanged}
      value={selectedValue}
      loading={isLoading}
      options={options as IGene[]}
      onInputChange={onChange}
      isOptionEqualToValue={() => true}
      filterOptions={((filterOptions) => filterOptions)}
      autoComplete
    />
  );
};

export default AutocompleteInput;
