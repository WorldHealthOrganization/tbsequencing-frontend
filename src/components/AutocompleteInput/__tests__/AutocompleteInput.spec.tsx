import React from 'react';
import {
  getTreeJSONWithRedux, render, screen, fireEvent,
} from '../../../utils/testUtils/jestUtils';
import AutocompleteInput from '../AutocompleteInput';
import { IGene } from '../../../services/genesApi/models';

const mockGene: IGene = {
  endPos: 0,
  id: 0,
  locusTag: 'RV00123',
  startPos: 0,
  strand: -1,
  geneName: 'rpoC',
  geneDbCrossrefId: 114,
};

const mockOptions: IGene[] = [
  mockGene,
  {
    endPos: 0,
    id: 0,
    locusTag: 'RV00124',
    startPos: 0,
    strand: -1,
    geneName: 'rpoB',
    geneDbCrossrefId: 118,
  },
];

describe('AutocompleteInput component test', () => {
  const mockOnChange = jest.fn();
  const mockOnSelectionChange = jest.fn();

  test('renders AutocompleteInput component without error', () => {
    render(
      <AutocompleteInput
        isLoading={false}
        onSelectionChanged={mockOnSelectionChange}
        onChange={mockOnChange}
        value="mock"
        selectedValue={mockGene}
        options={mockOptions}
        style={{}}
      />,
    );
  });

  test('handles onChange event', () => {
    render(<AutocompleteInput
      isLoading={false}
      onSelectionChanged={mockOnSelectionChange}
      onChange={mockOnChange}
      value="RV00123"
      selectedValue={mockGene}
      options={mockOptions}
      style={{}}
    />);

    const input = screen.getByDisplayValue('RV00123');

    fireEvent.change(input, { target: { value: '123' } });

    expect(mockOnChange).toHaveBeenCalled();
  });

  test('handles onSelectionChange event', async () => {
    render(<AutocompleteInput
      isLoading={false}
      onSelectionChanged={mockOnSelectionChange}
      onChange={mockOnChange}
      value="RV00"
      selectedValue={mockGene}
      options={mockOptions}
      style={{}}
    />);

    const container = screen.getByTestId('autocomplete');
    fireEvent.click(container);
    fireEvent.focus(container);

    const input = screen.getByDisplayValue('RV00123');

    fireEvent.change(input, { target: { value: 'RV00124' } });

    const dropDownOption = await screen.findByText('RV00124');

    fireEvent.click(dropDownOption);

    expect(mockOnSelectionChange).toHaveBeenCalled();
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <AutocompleteInput
        isLoading={false}
        onSelectionChanged={mockOnSelectionChange}
        onChange={mockOnChange}
        value="mock"
        selectedValue={mockGene}
        options={[mockGene]}
        style={{}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});
