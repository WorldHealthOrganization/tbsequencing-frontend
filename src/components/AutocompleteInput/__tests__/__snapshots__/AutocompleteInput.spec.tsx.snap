// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteInput component test Should render correctly 1`] = `
<div
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
  data-testid="autocomplete"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-expanded={false}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        placeholder="Search for gene..."
        readOnly={false}
        required={false}
        role="combobox"
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Clear"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Clear"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-hdw1oc"
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
